sdk/core/azure-core/src/main/java/com/azure/core/util/AuthorizationChallengeHandler.java

```diff
@@ -0,0 +1,521 @@
+    }
+
+    /**
+     * Handles Basic authentication challenges.
+     *
+     * @return Authorization header for Basic authentication challenges.
+     */
+    public final String handleBasic() {
+        authorizationPipeliningType.set(BASIC);
+        String token = username + ":" + password;
+        return BASIC + Base64.getEncoder().encodeToString(token.getBytes(StandardCharsets.UTF_8));
+    }
+
+    /**
+     * Handles Digest authentication challenges.
+     *
+     * @param method HTTP method being used in the request.
+     * @param uri Relative URI for the request.
+     * @param challenges List of challenges that the server returned for the client to choose from and use when creating
+     * the authorization header.
+     * @param entityBodySupplier Supplies the request entity body, used to compute the hash of the body when using
+     * {@code "qop=auth-int"}.
+     * @return Authorization header for Digest authentication challenges.
+     */
+    public final String handleDigest(String method, String uri, List<Map<String, String>> challenges,
+        Supplier<byte[]> entityBodySupplier) {
+        authorizationPipeliningType.set(DIGEST);
+        Map<String, List<Map<String, String>>> challengesByType = partitionByChallengeType(challenges);
+
+        for (String algorithm : ALGORITHM_PREFERENCE_ORDER) {
+            // No challenges using this algorithm, skip it.
+            if (!challengesByType.containsKey(algorithm)) {
+                continue;
+            }
+
+            Function<byte[], byte[]> digestFunction = getDigestFunction(algorithm);
+
+            // Unable to retrieve a digest for the specified algorithm, skip it.
+            if (digestFunction == null) {
+                continue;
+            }
+
+            Map<String, String> challenge = new ConcurrentHashMap<>(challengesByType.get(algorithm).get(0));
+            lastChallenge.set(challenge);
+
+            return createDigestAuthorizationHeader(method, uri, challenge, algorithm, entityBodySupplier,
+                digestFunction);
+        }
+
+        return null;
+    }
+
+    /**
+     * Attempts to pipeline requests by applying the most recent authorization type used to create an authorization
+     * header.
+     *
+     * @param method HTTP method being used in the request.
+     * @param uri Relative URI for the request.
+     * @param entityBodySupplier Supplies the request entity body, used to compute the hash of the body when using
+     * {@code "qop=auth-int"}.
+     * @return A preemptive authorization header for a potential Digest authentication challenge.
+     */
+    public final String attemptToPipelineAuthorization(String method, String uri, Supplier<byte[]> entityBodySupplier) {
+        String pipeliningType = authorizationPipeliningType.get();
+
+        if (DIGEST.equals(pipeliningType)) {
+            Map<String, String> challenge = lastChallenge.get();
+            String algorithm = challenge.get(ALGORITHM);
```

Comment:
I guess there might be a race condition here - `challenge.set()` is happening in `handleDigest()` and `challenge.get()` is being called here. Multiple threads calling these two methods in random order can lead to unexpected results unless there is some thread-safety built outside of this class that I am missing.

Reply:
There isn't thread safety built-in for this, should accesses to `lastChallenge` be synchronized?

Reply:
Synchronized access to `lastChallenge`